### PR TITLE
lightning: specify key range when switch mode

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: 
close https://github.com/pingcap/tidb/issues/45481
ref https://github.com/tikv/tikv/issues/14916 

Problem Summary:

### What is changed and how it works?
- set table range when switch tikv mode
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/cmd/tidb-lightning-ctl/main.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func run() error {
 		return errors.Trace(fetchMode(ctx, cfg, tls))
 	}
 	if len(*mode) != 0 {
-		return errors.Trace(lightning.SwitchMode(ctx, cfg, tls, *mode))
+		return errors.Trace(lightning.SwitchMode(ctx, cfg, tls, *mode, nil))
```
Comment (lance6716) [inline br/cmd/tidb-lightning-ctl/main.go:102]: not sure multi-rocks will change import mode for all regions when it's not specified

File: br/pkg/lightning/importer/import.go [hunk 2]
```diff
@@ -1177,6 +1178,7 @@ func (rc *Controller) buildRunPeriodicActionAndCancelFunc(ctx context.Context, s
 		cancelFuncs = append(cancelFuncs, func(bool) { switchModeTicker.Stop() })
 		cancelFuncs = append(cancelFuncs, func(do bool) {
 			if do {
+				rc.tikvModeSwitcher.SetKeyRanges(rc.buildTablesRanges())
```
Comment (lance6716) [inline br/pkg/lightning/importer/import.go:1181]: will it change? Why do we need to set it again

File: ddl/util/util.go [hunk 3]
```diff
@@ -325,3 +326,41 @@ func WrapKey2String(key []byte) string {
...
+const (
+	getRaftKvVersionSQL = "show config where type = 'tikv' && name = 'storage.engine'"
+	raftKv2             = "raft-kv2"
+)
+
+// IsRaftKv2 checks whether the raft-kv2 is enabled
+func IsRaftKv2(ctx context.Context, sctx sessionctx.Context) (bool, error) {
+	// Mock store does not support `show config` now, so we  use failpoint here
+	// to control whether we are in raft-kv2
+	failpoint.Inject("isRaftKv2", func(v failpoint.Value) (bool, error) {
```
Comment (lance6716) [inline ddl/util/util.go:339]: maybe intest.InTest

File: ddl/util/util.go [hunk 4]
```diff
@@ -325,3 +326,41 @@ func WrapKey2String(key []byte) string {
...
+	rs, err := sctx.(sqlexec.SQLExecutor).ExecuteInternal(ctx, getRaftKvVersionSQL)
+	if rs != nil {
+		defer terror.Call(rs.Close)
+	}
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	req := rs.NewChunk(nil)
+	it := chunk.NewIterator4Chunk(req)
+	err = rs.Next(context.TODO(), req)
```
Comment (lance6716) [inline ddl/util/util.go:353]: we can use the context in argument

File: ddl/util/util.go [hunk 5]
```diff
@@ -325,3 +326,41 @@ func WrapKey2String(key []byte) string {
...
+	})
+
+	rs, err := sctx.(sqlexec.SQLExecutor).ExecuteInternal(ctx, getRaftKvVersionSQL)
+	if rs != nil {
+		defer terror.Call(rs.Close)
+	}
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	req := rs.NewChunk(nil)
```
Comment (lance6716) [inline ddl/util/util.go:351]: maybe DrainRecordSet

File: disttask/importinto/dispatcher.go [hunk 6]
```diff
@@ -350,7 +350,7 @@ func (h *flowHandle) updateCurrentTask(task *proto.Task) {
 	if h.currTaskID.Swap(task.ID) != task.ID {
 		taskMeta := &TaskMeta{}
 		if err := json.Unmarshal(task.Meta, taskMeta); err == nil {
-			h.disableTiKVImportMode.Store(taskMeta.Plan.DisableTiKVImportMode)
+			h.disableTiKVImportMode.Store(taskMeta.Plan.DisableTiKVImportMode || taskMeta.Plan.IsRaftKV2)
```
Comment (lance6716) [inline disttask/importinto/dispatcher.go:353]: why we will disable it?

File: ddl/util/util.go (hunk 3 above)
Comment (GMHDBJD) [reply ddl/util/util.go]: move from https://github.com/pingcap/tidb/blob/79f00016016e9a1688613b7f4c8b5d830814fecb/store/gcworker/gc_worker.go#L840
so no change the original implementation

File: ddl/util/util.go (hunk 4 above)
Comment (GMHDBJD) [reply ddl/util/util.go:353]: ditto

File: ddl/util/util.go (hunk 5 above)
Comment (GMHDBJD) [reply ddl/util/util.go:351]: ditto

File: disttask/importinto/dispatcher.go (hunk 6 above)
Comment (GMHDBJD) [reply disttask/importinto/dispatcher.go:353]: for raftkv2, switch mode in local backend

File: br/pkg/lightning/common/util.go [hunk 7]
```diff
@@ -648,3 +648,28 @@ func getSessionVariable(ctx context.Context, db *sql.DB, variable string) (value
 
 	return value, nil
 }
+
+// IsRaftKV2 checks whether the raft-kv2 is enabled
+func IsRaftKV2(ctx context.Context, db *sql.DB) (bool, error) {
+	var (
+		getRaftKvVersionSQL       = "show config where type = 'tikv' && name = 'storage.engine'"
```
Comment (D3Hunter) [inline br/pkg/lightning/common/util.go:655]: maybe use `and`, see https://stackoverflow.com/a/4105810

File: disttask/importinto/dispatcher.go (hunk 6 above)
Comment (D3Hunter) [reply disttask/importinto/dispatcher.go:353]: maybe leave a comment here

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: ddl/util/util.go (hunk 4 above)
Comment (GMHDBJD) [reply ddl/util/util.go]: [3806d90](https://github.com/pingcap/tidb/pull/45482/commits/3806d90b9350d6248f375a39a60f4b9d85e80d4c)

File: ddl/util/util.go (hunk 5 above)
Comment (GMHDBJD) [reply ddl/util/util.go]: [3806d90](https://github.com/pingcap/tidb/pull/45482/commits/3806d90b9350d6248f375a39a60f4b9d85e80d4c)

File: disttask/importinto/dispatcher.go (hunk 6 above)
Comment (GMHDBJD) [reply disttask/importinto/dispatcher.go:354]: [3806d90](https://github.com/pingcap/tidb/pull/45482/commits/3806d90b9350d6248f375a39a60f4b9d85e80d4c)

File: br/pkg/lightning/backend/local/local.go [hunk 8]
```diff
@@ -1695,6 +1715,51 @@ func (local *Backend) LocalWriter(_ context.Context, cfg *backend.LocalWriterCon
 	return openLocalWriter(cfg, engine, local.tikvCodec, local.LocalWriterMemCacheSize, local.bufferPool.NewBuffer())
 }
 
+// SwitchModeByKeyRanges will swith tikv mode for regions in the specific key range for multirocksdb.
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/local.go:1718]:
```suggestion
// SwitchModeByKeyRanges will switch tikv mode for regions in the specific key range for multirocksdb.
```

File: ddl/util/util.go [hunk 9]
```diff
@@ -325,3 +326,39 @@ func WrapKey2String(key []byte) string {
...
+	rows, err := sqlexec.DrainRecordSet(ctx, rs, sctx.GetSessionVars().MaxChunkSize)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if len(rows) == 0 {
+		return false, nil
+	}
+
+	// All nodes should have the same type of engine
+	raftVersion := rows[0].GetString(3)
```
Comment (lance6716) [inline ddl/util/util.go:362]: In future we can add a test to make sure the column "3" is not changed.

**Pull request merged**